### DISCUSSION TITLE
Busy-wait inside cudaDeviceSynchronize while calling Thrust algorithms

### DISCUSSION THREAD

**Discussion by andrewmcdonald-oxb** [NVIDIA/cccl#1053] [Thrust] [answered]
Hello. Not sure if this is the right place to ask a question. But I've inherited an app to profile on x64 Linux, which makes use of Thrust (specifically `exclusive_scan_n` in this case). The profile shows it spends a lot of CPU time busy-waiting inside `cudaDeviceSynchronize`.

Is there some way of telling Thrust to sleep the CPU thread instead while waiting for the result from the GPU? I've read the docs I could find but couldn't spot anything about making this choice. Does Thrust just use the default Cuda stream? Perhaps there's a way to set a flag globally on that? I'm familiar with GPU programming but haven't much experience with Cuda yet, so I might be missing something.

**Comment by alliepiper** [comment, marked as answer]
The wait behavior is controlled by `cudaSetDeviceFlags` ([docs](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g69e73c7dda3fc05306ae7c811a690fac)). The `cudaDeviceScheduleYield` should do what you're looking for.

Thrust uses the default stream by default, but this can be changed by passing `thrust::device.on(stream)` as the first argument to the algorithm, where `stream` is a `cudaStream_t`.

If you want more control over synchronization, you may also want to look at `cub::DeviceScan::ExclusiveScan` ([docs](https://nvlabs.github.io/cub/structcub_1_1_device_scan.html)). While Thrust algorithms are inherently synchronous, the CUB algorithms provide the user with more flexibility in this regard.

**Reply by andrewmcdonald-oxb** [reply]
Thanks Allison, that's helpful! Setting the flag reduces the CPU load for our TensorRT calls as well due to their internals.

I'm opting for the `BlockingSync` flag at the moment; does the `Yield` one re-yield if the scheduler picks the thread back up again and the operation still hasn't finished? I guess we're getting away from Thrust now though.

**Reply by alliepiper** [reply]
I would assume so, but I'm not sure. You might want to ask about that one on the CUDA dev forums.
